### PR TITLE
Add flag to masquerade all in kube-proxy when using iptables proxier

### LINKED ISSUE #12843: kube-proxy: iptables implementation: allow a fallback flag to MASQUERADE pod source IPs
Related to #3760 .

As a safety mesure, it could be very useful to actually include a flag to proxy to MASQUERADE everything (acting very close to a userspace proxy). I think it's not very hard to do and a very good way to diagnose or even fallback in case of problem (I'm thinking about vxlan cases).

Something like "--snat-clients" or "--snat-pods" could be appropriate.

### PR DESCRIPTION
**Pull request by BenTheElder**
This closes https://github.com/kubernetes/kubernetes/issues/12843

### CODE REVIEW

**Comment by MikaelCluseau** [comment]
LGTM, thanks :-)
Maybe some test coverage (e2e?) will be required, though.

**Comment by BenTheElder** [comment]
Yw! Just doing a few more little things before GSoC ends.

I think in general we need a lot more e2e. It's on @thockin's list of
kube-proxy todos. (discussion here:
https://github.com/kubernetes/kubernetes/pull/12764#issuecomment-132264192)

**Comment by thockin** [comment]
Yeah e2e is on me before we can enable this new proxier.  Time ran out for
GSOC :(

File: cmd/kube-proxy/app/server.go [hunk 1]
```diff
@@ -61,6 +61,7 @@ type ProxyServer struct {
 	ForceUserspaceProxy bool
 	SyncPeriod          time.Duration
 	nodeRef             *api.ObjectReference // Reference to this node.
+	SnatAll             bool
```
Comment (thockin) [inline cmd/kube-proxy/app/server.go]: don't mix-case acronyms in Go.

MasqueradeAll or SNATAll.

File: cmd/kube-proxy/app/server.go [hunk 2]
```diff
@@ -88,6 +89,7 @@ func (s *ProxyServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.HostnameOverride, "hostname-override", s.HostnameOverride, "If non-empty, will use this string as identification instead of the actual hostname.")
 	fs.BoolVar(&s.ForceUserspaceProxy, "legacy-userspace-proxy", true, "Use the legacy userspace proxy (instead of the pure iptables proxy).")
 	fs.DurationVar(&s.SyncPeriod, "iptables-sync-period", 5*time.Second, "How often iptables rules are refreshed (e.g. '5s', '1m', '2h22m').  Must be greater than 0.")
+	fs.BoolVar(&s.SnatAll, "snat-all", false, "If using the pure iptables proxy, SNAT everything")
```
Comment (thockin) [inline cmd/kube-proxy/app/server.go]: I prefer masquerade-all to snat-all - it's more precise.

**Comment by thockin** [comment]
Naming, then LGTM

**Comment by MikaelCluseau** [comment]
On 08/21/2015 09:24 AM, Tim Hockin wrote:
> Yeah e2e is on me before we can enable this new proxier. Time ran out for
> GSOC :(

I'll try to help, but I haven't coded much in/around Kubernetes. I'm 
prototyping some things interacting with Kubernetes to get at a 
reasonable level.

**Comment by thockin** [comment]
Any help is appreciated!  e2e is tedious but SO SO important.  I have a
sketch of the interesting cases we need to make sure are covered

**Comment by MikaelCluseau** [comment]
Okay, I'll pick one, or you can assign me one to begin with. I'll probably only have week-end time on this, though.

**Comment by BenTheElder** [comment]
Renamed everything to `MasqueradeAll` since it's more accurate.

**Comment by thockin** [comment]
LGTM

**Comment by BenTheElder** [comment]
Will squash in a sec.

**Comment by BenTheElder** [comment]
Gotta fix the known-flags file again. Will fixup.

**Comment by BenTheElder** [comment]
Should be fixed.

**Comment by thockin** [comment]
Still lgtm

**Comment by BenTheElder** [comment]
For e2e: Looks like all the tests passed but the build timed out?

**Comment by thockin** [comment]
@k8s-bot test this please

**Comment by BenTheElder** [comment]
huh.
```
Summarizing 2 Failures:

[Fail] Services [It] should serve multiport endpoints from pods 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:1126

[Fail] Services [It] should serve a basic endpoint from pods 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:1126

Ran 84 of 146 Specs in 583.312 seconds
FAIL! -- 82 Passed | 2 Failed | 2 Pending | 60 Skipped 
```

**Comment by thockin** [comment]
If you run it by hand - does it pass?

**Comment by thockin** [comment]
@k8s-bot test this please

**Comment by BenTheElder** [comment]
Not sure yet, I didn't expect e2e to be effected, it should still be
running the userspace proxy (?)

**Comment by thockin** [comment]
I kicked the tester again.  let's see...

**Pull request merged**